contracts/NFTPool.sol

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.7.5;
+pragma experimental ABIEncoderV2;
+
+import {DaiPool, Erc20Pool, ReceiverWeight, Dai} from "./Pool.sol";
+import {IERC721} from "openzeppelin-contracts/token/ERC721/IERC721.sol";
+
+/// @notice NFT pool contract to support streaming based on NFT ownership
+/// A NFT can be a sender or a receiver, a unique id is generated based on
+/// NFT registry address and the tokenId
+contract NFTPool is Erc20Pool {
+    modifier nftOwner (address nftRegistry, uint tokenId) {
+        require(IERC721(nftRegistry).ownerOf(tokenId) == msg.sender, "not-NFT-owner"); _;
+    }
+    constructor(uint64 cycleSecs, Dai dai) Erc20Pool(cycleSecs, dai) {}
+
+    /// @notice generates a unique 20 bytes by hashing the nft registry  and tokenId
+    /// @param nftRegistry address of the NFT specific registry
+    /// @param tokenId the unique token id for the NFT registry
+    function nftID(address nftRegistry, uint128 tokenId) public pure returns (address id) {
+        // gas optimized without local variables
```

Comment:
I didn't know that trick! What does it do? Does it skip memory allocation when passing data from `encodePacked` to `keccak256`?

BTW integer casting of local variables in Solidity is a noop in EVM, they are always uint256, it only changes how they're handled by Solidity :)

Reply:
Ah no, I just added to comment because in a normal programming language you might would make some local variables to increase readability. 

```
bytes32 hash = ...;
address id = ...;
return id;
```

Yeah, the casting looks a bit ugly but you can't cast bytes32 to uint160 directly. We might can do it in  `assembly`

Reply:
Casting is fine, I just thought that the optimizer is smart enough to not create the unneeded local variables. That would leave us more freedom with formatting, as you've shown.